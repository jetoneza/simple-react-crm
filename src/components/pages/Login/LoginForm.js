import React from 'react';
import PropTypes from 'prop-types';
import Validator from 'validator';
import { Form } from 'semantic-ui-react';

class LoginForm extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
  };

  state = {
    data: {
      email: '',
      password: '',
    },
    errors: {},
  };

  handleSubmit = (e) => {
    e.preventDefault();

    const errors = this.validate(this.state.data);

    this.setState({ errors });

    if (Object.keys(errors).length !== 0) {
      return;
    }

    this.props.onSubmit(this.state.data);
  };

  validate = (data) => {
    const errors = {};

    if (!data.password) {
      errors.password = 'Please provide a password.';
    }

    if (!Validator.isEmail(data.email)) {
      errors.email = 'Invalid email.';
    }

    return errors;
  };

  handleOnChange = (e) => {
    const { name, value } = e.target;

    this.setState({
      data: {
        ...this.state.data,
        [name]: value,
      },
    });
  };

  renderError = (error) => {
    return error && <span style={{ color: 'red' }}>{ error }</span>;
  };

  render() {
    const { data, errors } = this.state;

    return (
      <Form onSubmit={this.handleSubmit}>
        <div className="ui stacked segment">
          <Form.Field error={!!errors.email}>
            <div className="ui left icon input">
              <i className="user icon"></i>
              <input
                type="email"
                name="email"
                placeholder="E-mail address"
                onChange={this.handleOnChange}
                value={data.email}/>
            </div>
            { this.renderError(errors.email) }
          </Form.Field>
          <Form.Field error={!!errors.password}>
            <div className="ui left icon input">
              <i className="lock icon"></i>
              <input
                type="password"
                name="password"
                placeholder="Password"
                onChange={this.handleOnChange}
                value={data.password}/>
            </div>
            { this.renderError(errors.password) }
          </Form.Field>
          <button className="ui fluid large green submit button">Login</button>
        </div>
      </Form>
    );
  }
}

export default LoginForm;
