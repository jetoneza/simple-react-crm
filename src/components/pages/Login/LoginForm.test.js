import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import { mount } from 'enzyme';

import LoginForm from './LoginForm';

describe('<LoginForm /> test', () => {
  const onSubmit = jest.fn();

  test('Should display errors if exists', () => {
    const wrapper = mount(<LoginForm onSubmit={onSubmit} />);

    // No errors should be displayed at first
    expect(wrapper.find('span.error').length).toEqual(0);

    wrapper.setState({
      errors: {
        password: 'has password error',
        email: 'email has error',
      },
    });

    expect(wrapper.find('span.error').length).toEqual(2);
  });

  test('Should update state on input change', () => {
    const wrapper = mount(<LoginForm onSubmit={onSubmit} />);

    const input = wrapper.find('input').first();

    input.simulate('change', { target: { name: 'email', value: 'test@email.com' } });
    input.simulate('change', { target: { name: 'password', value: '1234' } });

    expect(wrapper.state().data.email).toEqual('test@email.com');
    expect(wrapper.state().data.password).toEqual('1234');
  });

  test('Should call onSubmit prop if form data is valid', () => {
    const onSubmitCallback = sinon.spy();
    const wrapper = mount(<LoginForm onSubmit={onSubmitCallback} />);
    wrapper.setState({
      data: {
        email: 'test@email.com',
        password: '1234',
      },
    })

    const form = wrapper.find('form');
    form.simulate('submit');

    expect(onSubmitCallback.called).toBeTruthy();
  });

  test('Should not call onSubmit prop if form data invalid', () => {
    const onSubmitCallback = sinon.spy();
    const wrapper = mount(<LoginForm onSubmit={onSubmitCallback} />);
    wrapper.setState({
      data: {
        email: 'test@email',
        password: '',
      },
    })

    const form = wrapper.find('form');
    form.simulate('submit');

    expect(wrapper.state().errors.email).not.toBeNull();
    expect(wrapper.state().errors.password).not.toBeNull();
    expect(onSubmitCallback.called).toBeFalsy();
  });

  test('Should render login form', () => {
    const component = <LoginForm onSubmit={onSubmit} />;
    const wrapper = mount(component);

    expect(wrapper.find('form').length).toEqual(1);
    expect(wrapper.find('input').length).toEqual(2);
    expect(wrapper.find('button').length).toEqual(1);

    const renderedComponent = renderer.create(component);
    const tree = renderedComponent.toJSON();

    expect(tree).toMatchSnapshot();
  });
});

