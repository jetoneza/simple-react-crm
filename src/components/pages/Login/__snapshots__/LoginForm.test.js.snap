// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> test Should render login form 1`] = `
<form
  action={undefined}
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="ui stacked segment"
  >
    <div
      className="field"
    >
      <div
        className="ui left icon input"
      >
        <i
          className="user icon"
        />
        <input
          name="email"
          onChange={[Function]}
          placeholder="E-mail address"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <div
        className="ui left icon input"
      >
        <i
          className="lock icon"
        />
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="ui fluid large green submit button"
    >
      Login
    </button>
  </div>
</form>
`;
